pkg/update/update_cockroach_version_common.go

```diff
@@ -46,39 +48,66 @@ func makeUpdateCockroachVersionFunction(
+// We need to return a status and an error instead of just an error.
+
 // makeWaitUntilCRDBPodIsRunningNewVersionFunction takes a cockroachImage and
 // returns a function which takes a Kubernetes clientset, a statefulset, and a
 // pod number within the statefulset. This function checks if the specified
 // pod is running the new cockroachImage version and is in a `ready` state.
 func makeIsCRBPodIsRunningNewVersionFunction(
 	cockroachImage string,
-) func(update *UpdateSts, podNumber int, l *zap.Logger) error {
-	return func(update *UpdateSts, podNumber int, l *zap.Logger) error {
+) func(update *UpdateSts, podNumber int, l logr.Logger) error {
+	return func(update *UpdateSts, podNumber int, l logr.Logger) error {
 		sts := update.sts
 		stsName := sts.Name
 		stsNamespace := sts.Namespace
 		podName := fmt.Sprintf("%s-%d", stsName, podNumber)
 		clientset := update.clientset
-		crdbPod, err := clientset.CoreV1().Pods(stsNamespace).Get(context.TODO(), stsName, metav1.GetOptions{})
-		if err != nil {
-			return errors.Wrapf(err, "error getting cockroach pod")
+
+		crdbPod, err := clientset.CoreV1().Pods(stsNamespace).Get(update.ctx, podName, metav1.GetOptions{})
+		if k8sErrors.IsNotFound(err) { // this is not an error
+			l.Info("cannot find Pod", "podName", podName, "namespace", stsNamespace)
+			return err
+		} else if statusError, isStatus := err.(*k8sErrors.StatusError); isStatus { // this is an error
+			l.Error(statusError, fmt.Sprintf("status error getting pod %v", statusError.ErrStatus.Message))
+			return statusError
+		} else if err != nil { // this is an error
+			l.Error(err, "error getting pod")
+			return err
 		}
+
 		for i := range crdbPod.Spec.Containers {
 			container := &crdbPod.Spec.Containers[i]
-			if container.Name == "cockroachdb" {
+			// TODO this is hard coded and resource statefulset needs to use this
+			if container.Name == "db" {
+
+				// TODO this is not an error but should return a wait status
 				if container.Image != cockroachImage {
+					l.Info("Pod is not updated to current image.")
 					return fmt.Errorf("%s pod is on image %s, expected %s", podName, container.Image, cockroachImage)
 				}
+
+				// TODO this is not an error but should return a wait status
 				// CRDB pod is updated to new Cockroach image. Now check
 				// that the pod is in a ready state before proceeding.
 				if !IsPodReady(crdbPod) {
-					return fmt.Errorf("%s pod not ready yet", podName)
+					l.Info("Pod is not ready yet.", "pod name", crdbPod)
+					return fmt.Errorf("%s pod not ready yet", crdbPod)
 				}
-				l.Sugar().Infof("%s is running new version on %s", podName, stsNamespace)
+
+				l.Info("is running new version on", "podName", podName, "stsName", stsNamespace)
 				return nil
 			}
 		}
-		return fmt.Errorf("cockroachdb container not found within the cockroach pod")
+
+		// TODO how do we even get here??
+		// I am not certain that this code is even used.
```

Comment:
Looks like a sanity check?

Might be more reasonable/understandable to do something like:

```
var container *TheType
for i, c := range pod.Spec.Containers {
    if c.Name == "..." {
        container = &pod.Spec.Containers[i]
    }
}

if container == nil {
  return fmt.Errorf("how did this happen?")
}

// rest of code ...
```